src/Neo/SmartContract/Native/Notary.cs

```diff
@@ -0,0 +1,343 @@
+using System;
+using System.Linq;
+using System.Numerics;
+using Array = Neo.VM.Types.Array;
+
+namespace Neo.SmartContract.Native
+{
+    /// <summary>
+    /// The Notary native contract used for multisignature transactions forming assistance.
+    /// </summary>
+    public sealed class Notary : NativeContract
+    {
+        /// <summary>
+        /// A default value for maximum allowed NotValidBeforeDelta. It is set to be
+        /// 20 rounds for 7 validators, a little more than half an hour for 15-seconds blocks.
+        /// </summary>
+        private const int DefaultMaxNotValidBeforeDelta = 140;
+        /// <summary>
+        /// A default value for deposit lock period.
+        /// </summary>
+        private const int DefaultDepositDeltaTill = 5760;
+        private const byte Prefix_Deposit = 1;
+        private const byte Prefix_MaxNotValidBeforeDelta = 10;
+
+        internal Notary() : base() { }
+
+        public override Hardfork? ActiveIn => Hardfork.HF_Echidna;
+
+        internal override ContractTask InitializeAsync(ApplicationEngine engine, Hardfork? hardfork)
+        {
+            if (hardfork == ActiveIn)
+            {
+                engine.SnapshotCache.Add(CreateStorageKey(Prefix_MaxNotValidBeforeDelta), new StorageItem(DefaultMaxNotValidBeforeDelta));
+            }
+            return ContractTask.CompletedTask;
+        }
+
+        internal override async ContractTask OnPersistAsync(ApplicationEngine engine)
+        {
+            long nFees = 0;
+            ECPoint[]? notaries = null;
+            foreach (var tx in engine.PersistingBlock.Transactions)
+            {
+                var attr = tx.GetAttribute<NotaryAssisted>();
+                if (attr is not null)
+                {
+                    notaries ??= GetNotaryNodes(engine.SnapshotCache);
+                    var nKeys = attr.NKeys;
+                    nFees += (long)nKeys + 1;
+                    if (tx.Sender == Hash)
+                    {
+                        var payer = tx.Signers[1];
+                        // Don't need to seal because Deposit is a fixed-sized interoperable, hence always can be serialized.
+                        var balance = engine.SnapshotCache.GetAndChange(CreateStorageKey(Prefix_Deposit, payer.Account))?.GetInteroperable<Deposit>();
+                        if (balance != null)
+                        {
+                            balance.Amount -= tx.SystemFee + tx.NetworkFee;
+                            if (balance.Amount.Sign == 0) RemoveDepositFor(engine.SnapshotCache, payer.Account);
+                        }
+                    }
+                }
+            }
+            if (nFees == 0) return;
+            if (notaries == null) return;
+            var singleReward = CalculateNotaryReward(engine.SnapshotCache, nFees, notaries.Length);
+            foreach (var notary in notaries) await GAS.Mint(engine, Contract.CreateSignatureRedeemScript(notary).ToScriptHash(), singleReward, false);
```

Comment:
Are we sure that this matches ` totalNetworkFee -= (notaryAssisted.NKeys + 1) * Policy.GetAttributeFeeV1(engine.SnapshotCache, (byte)notaryAssisted.Type);`

Reply:
Yes, it is provided by NotaryAssisted attribute verification rules.

Reply:
See https://github.com/neo-project/neo/pull/3178#discussion_r2055707380.